src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs

```diff
@@ -0,0 +1,218 @@
+        }
+
+        public static IReadOnlyDictionary<Type, Type> CreateProxyTypeMap(RuntimeType typeMapGroup)
+        {
+            RuntimeTypeHandle typeMapGroupHandle = typeMapGroup.TypeHandle;
+            foreach (TypeManagerHandle module in RuntimeAugments.GetLoadedModules())
+            {
+                if (!TryGetNativeReaderForBlob(module, ReflectionMapBlob.AssociatedTypeMap, out NativeReader externalTypeMapReader))
+                {
+                    continue;
+                }
+                NativeParser externalTypeMapParser = new NativeParser(externalTypeMapReader, 0);
+                NativeHashtable externalTypeMapTable = new NativeHashtable(externalTypeMapParser);
+
+                ExternalReferencesTable externalReferences = default;
+                externalReferences.InitializeCommonFixupsTable(module);
+
+                var lookup = externalTypeMapTable.Lookup(typeMapGroupHandle.GetHashCode());
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    RuntimeTypeHandle foundTypeMapGroup = externalReferences.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                    if (!foundTypeMapGroup.Equals(typeMapGroupHandle))
+                    {
+                        continue;
+                    }
+                    bool isValid = entryParser.GetUnsigned() == 1;
+                    if (!isValid)
+                    {
+                        unsafe
+                        {
+                            delegate*<void> exceptionStub = (delegate*<void>)externalReferences.GetFunctionPointerFromIndex(entryParser.GetUnsigned());
+                            exceptionStub();
+                            Debug.Fail("Expected exception stub to throw an exception.");
+                            return null; // Should never reach here, as the exception stub should throw an exception.
+                        }
+                    }
+
+                    return new AssociatedTypeMapDictionary(new NativeHashtable(entryParser), externalReferences);
+                }
+            }
+
+            return new Dictionary<Type, Type>(); // Return an empty dictionary if no valid type map is found.
+        }
+
+        private static unsafe bool TryGetNativeReaderForBlob(TypeManagerHandle module, ReflectionMapBlob blob, out NativeReader reader)
+        {
+            byte* pBlob;
+            uint cbBlob;
+
+            if (RuntimeImports.RhFindBlob(module, (uint)blob, &pBlob, &cbBlob))
+            {
+                reader = new NativeReader(pBlob, cbBlob);
+                return true;
+            }
+
+            reader = default;
+            return false;
+        }
+
+        private class ExternalTypeMapDictionary(NativeHashtable table, ExternalReferencesTable externalReferences) : IReadOnlyDictionary<string, Type>
```

Comment:
Style: instance variables should be prefixed with underscores (not going to comment on all the places, this is not the only one).

Reply:
According to our coding style, primary constructor parameters should be named like parameters, even though they can be used throughout the body of a type.

However, that also says that I shouldn't use them when it's hard to figure out where they're coming from, so I'll adjust some of these to use fields for ease of reading outside of an IDE.

Reply:
I spoke to @stephentoub about the guidelines offline (and I need to put out a PR to the guidelines page with what he recommended).

Reply:
Depending on what the guidelines are, there might still be work to address here.

Reply:
Guidelines are now updated in main.

Reply:
> Guidelines are now updated in main.

Could you make a pass to ensure this PR matches the guideline? I know the guideline says "sufficiently small" and that term is fuzzy, but I tend to look at these as de minimis classes, around dozen lines where I don't need to keep clicking the expand button in GitHub UI when reviewing. This PR still uses it for 50-line classes and that's a lot of clicking to find where a variable comes from.